### DISCUSSION TITLE
Better C Generics Part 1: The Extendible _Generic

### DISCUSSION THREAD

**Discussion by JacksonAllan** [JacksonAllan/CC#1] [General]
Comments on the article [Better C Generics Part 1: The Extendible `_Generic`](https://github.com/JacksonAllan/CC/blob/main/articles/Better_C_Generics_Part_1_The_Extendible_Generic.md) belong below.

**Comment by g-berthiaume** [comment]
Fantastic article! 
I'm gonna go explore your codebase to better understand how you use this pattern.

btw @JacksonAllan is there a part 2 ?

**Reply by JacksonAllan** [reply]
Thanks for reading!

> I'm gonna go explore your codebase to better understand how you use this pattern.

It might be better to look at [Verstable](https://github.com/JacksonAllan/Verstable), CC’s sister project, first or instead. There, the pattern is used to provide a generic API (the `vt_` macros) for all the hash-table types created by the user. There are two issues with referring to CC as an example of the pattern:

* CC uses the pattern in a different way (namely to deduce hash, comparison, and destructor functions) that is rather limited in scope and fundamentally intertwined with other unusual techniques upon which the library relies. 
* CC relies on many novel techniques to achieve it goals. Hence, the codebase probably has a steep learning curve, even though everything is thoroughly documented with in-code comments. Verstable, on the other hand, is based on the conventional pseudo-template paradigm that many modern C container libraries use. It contains far less, and much more straightforward, code.

I probably should update the extendible-`_Generic` article to direct users to the Verstable example.

> is there a part 2 ?

Not yet, unfortunately. However, I recently published [this tangential article]( https://jacksonallan.github.io/c_cpp_hash_tables_benchmark/) benchmarking hash tables, including CC/Verstable.

The real part 2, when it eventually arrives, will be a comprehensive review of the different approaches to creating generic containers in C. Its final section will describe the way that CC extends what I called the “typed-pointer/hidden-metadata paradigm” in the README to accommodate containers with two datatypes (namely maps and, hopefully soon, ordered maps) and a compile-time integer identifying the container type. The crux of that mechanism is already described in my comment titled _Subverting C’s type system, i.e. associating extra compile-time information with pointers_ in [this Reddit thread]( https://www.reddit.com/r/C_Programming/comments/zvubfb/convenient_containers_a_usabilityoriented_generic/).

Part 3, if there is one, will detail the other tricks upon which CC relies, most of which are described in the aforementioned thread. When it comes to `_Generic`, the most relevant comment is probably the one I titled _A poor man's SFINAE_. The mechanism it describes was later described more fully by someone else [here](https://www.chiark.greenend.org.uk/~sgtatham/quasiblog/c11-generic/#coercion). Note, though, that the current version of CC no longer uses this technique (anywhere, as far as I recall) because I got better compile speed with a different solution to problem it solves.

By the way, CC’s [dev branch](https://github.com/JacksonAllan/CC/tree/dev) now includes MSVC support. I’ll be releasing it as a new version on the main branch soon.

**Reply by g-berthiaume** [reply]
> It might be better to look at [Verstable](https://github.com/JacksonAllan/Verstable)

Interesting, I will [check it out](https://github.com/JacksonAllan/Verstable/blob/099ebd6586b115646ef7b98e0e23de6a9792e01e/verstable.h#L598)!
 

>However, I recently published [this tangential article](https://jacksonallan.github.io/c_cpp_hash_tables_benchmark/) benchmarking hash tables, including CC/Verstable.
 
I've read it. I enjoyed the details and rigour; combining different load cases and benchmarks helped create interesting insight.

> The real part 2, when it eventually arrives, will be a comprehensive review of the different approaches to creating generic containers in C.

I can't wait to read it. IMO, good generic containers are the key to reconciliation type-safety and ergonomics in C.

I would be interested in seeing those approaches compared to code generation.
While reading the [raddebugger](https://github.com/EpicGamesExt/raddebugger) source code, I was surprised at how effective this technique seems to be (see the `metagen` section in the README).

> By the way, the CC’s [dev branch](https://github.com/JacksonAllan/CC/tree/dev) now includes MSVC support. I’ll be releasing it as a new version on the main branch soon. 

Your work is a joy to follow!

Regards,
Gabriel

**Reply by JacksonAllan** [reply]
Sorry for the delayed response.

> Your work is a joy to follow!

> I can't wait to read it. IMO, good generic containers are the key to reconciliation type-safety and ergonomics in C.

Thanks! I haven't started writing the second article yet, but it's coming up on the to-do list through which I'm trying to make at least a little progress every day.

> I would be interested in seeing those approaches compared to code generation.
While reading the [raddebugger](https://github.com/EpicGamesExt/raddebugger) source code, I was surprised at how effective this technique seems to be (see the metagen section in the README).

If we're willing to go down the custom-preprocessor route, then the possibilities are essentially endless. Remember, that's how C++ [started its life](https://en.wikipedia.org/wiki/Cfront) decades ago. The problem, as I see it, is adoption. CC offers significant improvement over other container libraries when it comes to ergonomics, yet it has gained only modest traction in the C community over the 1.5 years that it has existed (although I could certainly put more effort into promoting it). I think most C developers would be disinclined to adopt a generic-container tool that requires an extra build step.

**Reply by g-berthiaume** [reply]
Thanks for sharing @JacksonAllan  !
